mayastor/src/subsys/config.rs

```diff
@@ -0,0 +1,362 @@
+    /// default trait for that structure.
+    pub fn apply(&self) {
+        info!("Applying Mayastor configuration settings");
+        // note: nvmf target does not have a set method
+        self.nvme_bdev_opts.set();
+        self.bdev_opts.set();
+        self.iscsi_tgt_conf.set();
+    }
+
+    /// create any nexus bdevs any failure will be logged, but we will silently
+    /// continue and try to create as many as possible. Returns the number of
+    /// bdevs for which the creation failed.
+    async fn create_nexus_bdevs(&self) -> usize {
+        let mut failures = 0;
+        info!("creating nexus devices");
+
+        // we can't use iterators here as they are not async
+        if let Some(bdevs) = self.nexus_bdevs.as_ref() {
+            for nexus in bdevs {
+                info!("creating nexus {}", nexus.name);
+                match Byte::from_str(&nexus.size) {
+                    Ok(val) => {
+                        if nexus_create(
+                            &nexus.name,
+                            val.get_bytes() as u64,
+                            Some(&nexus.uuid),
+                            &nexus.children,
+                        )
+                        .await
+                        .is_err()
+                        {
+                            error!("Failed to create nexus {}", nexus.name);
+                            failures += 1;
+                        }
+                    }
+                    Err(_e) => {
+                        failures += 1;
+                        error!(
+                            "Invalid size {} for {}",
+                            &nexus.size, nexus.name
+                        );
+                    }
+                }
+            }
+        }
+        failures
+    }
+
+    /// create base bdevs and export these over nvmf if configured
+    async fn create_base_bdevs(&self) -> usize {
+        let mut failures: usize = 0;
+        if let Some(bdevs) = self.base_bdevs.as_ref() {
+            for bdev in bdevs {
+                info!("creating bdev {}", bdev.uri);
+                if bdev_create(&bdev.uri).await.is_err() {
+                    warn!(
+                        "failed to create bdev {} during config load",
+                        bdev.uri
+                    );
+                    failures += 1;
+                    continue;
+                }
+
+                let mut my_bdev = Bdev::lookup_by_name(&bdev.uri).unwrap();
+
+                // if we were given some UUID set it now
+                if let Some(uuid) = bdev.uuid.as_ref() {
+                    my_bdev.set_uuid(Some(uuid.clone()));
+                }
+
+                let uuid = my_bdev.uuid_as_string();
+                assert_eq!(bdev.uuid.as_ref(), Some(&uuid));
+
+                if !self.implicit_share_base {
+                    continue;
+                }
+
+                // share the bdev over nvmf
```

Comment:
Not supporting iscsi replicas via yaml, because we'll be nvmf-only long-term? If so, maybe that should be in the pull-request comments.

Reply:
@chriswldenyer absence of iSCSI does not imply anything in a particular right. As stated in the PR, this is an initial pass at a framework that we can extent easily and better suits our needs. (we can debate if it achieved that of course...)

The reason I did nvmf here is because we can already do iSCSI easily (with the older ini file) but for nvmf this was not yet possible in a way that would allow us to set various settings easily.

Reply:
Ok